src/BaseListView.php

```diff
@@ -4,248 +4,346 @@
-     */
-    protected $dataReader;
-    /**
-     * @var \Yiisoft\Data\Paginator\KeysetPaginator|\Yiisoft\Data\Paginator\OffsetPaginator|null
-     */
-    protected $paginator;
-    /**
-     * @var LinkPager|null You can use a different widget class by configuring the "class" element.
-     */
-    protected ?LinkPager $pager = null;
-    /**
-     * @var LinkSorter|null You can use a different widget class by configuring the "class" element.
-     */
+    protected PaginatorInterface $paginator;
     protected ?LinkSorter $sorter = null;
+    protected string $summary = 'Showing <b>{begin, number}-{end, number}</b> of <b>{totalCount, number}</b> ' .
+        '{totalCount, plural, one{item} other{items}}';
+    protected array $summaryOptions = [];
+    protected bool $showOnEmpty = false;
+    protected ?string $emptyText = 'No results found.';
+    protected bool $showEmptyText = true;
+    protected array $emptyTextOptions = ['class' => 'empty'];
+    protected string $layout = "{summary}\n{items}\n{pager}";
+    private const FRAMEWORKCSS = [
+        self::BOOTSTRAP,
+        self::BULMA,
+    ];
+    private string $frameworkCss = self::BOOTSTRAP;
+    private int $pageSize = 0;
+    private int $currentPage = 1;
+    private array $requestAttributes = [];
+    private array $requestQueryParams = [];
+    private TranslatorInterface $translator;
+
+    public function __construct(TranslatorInterface $translator)
+    {
+        $this->translator = $translator;
+    }
+
     /**
-     * @var string the HTML content to be displayed as the summary of the list view.
-     * If you do not want to show the summary, you may set it with an empty string.
-     * The following tokens will be replaced with the corresponding values:
-     * - `{begin}`: the starting row number (1-based) currently being displayed
-     * - `{end}`: the ending row number (1-based) currently being displayed
-     * - `{count}`: the number of rows currently being displayed
-     * - `{totalCount}`: the total number of rows available
-     * - `{page}`: the page number (1-based) current being displayed
-     * - `{pageCount}`: the number of pages available
-     */
-    protected string $summary = 'Showing <b>{begin, number}-{end, number}</b> of <b>{totalCount, number}</b> {totalCount, plural, one{item} other{items}}.';
-    /**
-     * @var array the HTML attributes for the summary of the list view.
-     * The "tag" element specifies the tag name of the summary element and defaults to "div".
+     * Renders the data models.
      *
-     * @see Html::renderTagAttributes() for details on how attributes are being rendered.
+     * @return string the rendering result.
      */
-    protected array $summaryOptions = ['class' => 'summary'];
+    abstract public function renderItems(): string;
+
+    protected function run(): string
+    {
+        if ($this->showOnEmpty || ($this->paginator->getTotalItems() > 0)) {
+            $content = preg_replace_callback(
```

Comment:
What does it do?

Reply:
https://github.com/yiisoft/yii-dataview/blob/91d2a6f1d133b30c7608a41b33feeb9428fbc47d/src/BaseListView.php#L188

Renders the tags  tokens defined in template, the code was already there.